languageservice/src/code-actions/quickfix/add-missing-inputs.ts

```diff
@@ -0,0 +1,67 @@
+import {CodeAction, TextEdit} from "vscode-languageserver-types";
+import {CodeActionProvider} from "../types";
+import {DiagnosticCode, MissingInputsDiagnosticData} from "../../validate-action";
+
+export const addMissingInputsProvider: CodeActionProvider = {
+  diagnosticCodes: [DiagnosticCode.MissingRequiredInputs],
+
+  createCodeAction(context, diagnostic): CodeAction | undefined {
+    const data = diagnostic.data as MissingInputsDiagnosticData | undefined;
+    if (!data) {
+      return undefined;
+    }
+
+    const edits = createInputEdits(data);
+    if (!edits) {
+      return undefined;
+    }
+
+    const inputNames = data.missingInputs.map(i => i.name).join(", ");
+
+    return {
+      title: `Add missing input${data.missingInputs.length > 1 ? "s" : ""}: ${inputNames}`,
+      edit: {
+        changes: {
+          [context.uri]: edits
+        }
+      }
+    };
+  }
+};
+
+function createInputEdits(data: MissingInputsDiagnosticData): TextEdit[] | undefined {
+  const edits: TextEdit[] = [];
+
+  if (data.hasWithKey && data.withIndent !== undefined) {
+    // `with:` exists - use its indentation + 2 for inputs
+    const inputIndent = " ".repeat(data.withIndent + 2);
```

Comment:
Instead of assuming `2`, have you considered matching the customer's indent? You could detect this based on the intent amount of the first job. For example:

```typescript
function detectIndentSize(rootToken: MappingToken): number {
  const jobsValue = rootToken.find("jobs");
  if (jobsValue && isMapping(jobsValue) && jobsValue.count > 0) {
    const firstJob = jobsValue.get(0);
    if (firstJob?.key.range && jobsValue.range) {
      return firstJob.key.range.start.column - jobsValue.range.start.column;
    }
  }
  return 2; // fallback
}
```

Reply:
Done, thank you ❤️ I wonder if indent size is something we wanna carry around with `WorkflowTemplate` itself, I suspect it'll keep coming in use 🤔